Unify run_z_probe

AnHardt opened the pull request:
Unify run_z_probe
Implicates double touch for DELTAs when probing.
Introduce Z_PROBE_SPEED_FAST and Z_PROBE_SPEED_SLOW,
  defaulting to homing_feedrate_mm_m[Z_AXIS] and homing_feedrate_mm_m[Z_AXIS]/2
#4261

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2475,7 +2424,7 @@ static void homeaxis(AxisEnum axis) {
         }
       #endif
-      do_blocking_move_to_axis_pos(axis, endstop_adj[axis], set_homing_bump_feedrate(axis));
+      do_blocking_move_to_axis_pos(axis, endstop_adj[axis], get_homing_bump_feedrate(axis));
```
Looks like now this should be…

``` cpp
line_to_axis_pos(axis, endstop_adj[axis], get_homing_bump_feedrate(axis));
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2437,7 +2386,7 @@ static void homeaxis(AxisEnum axis) {
 
   // Move slowly towards the endstop until triggered
-  do_blocking_move_to_axis_pos(axis, 2 * home_bump_mm(axis) * axis_home_dir, set_homing_bump_feedrate(axis));
+  do_blocking_move_to_axis_pos(axis, 2 * home_bump_mm(axis) * axis_home_dir, get_homing_bump_feedrate(axis));
```
Looks like now this should be…

``` cpp
line_to_axis_pos(axis, 2 * home_bump_mm(axis) * axis_home_dir, get_homing_bump_feedrate(axis));
```

thinkyhead commented:
I rebased this on the latest code, then decided to make another tweak. Namely, it seems best not to include `homing_feedrate_mm_m[Z_AXIS]` as a default configuration option value, but to use a "factor" instead, which is multiplied by the homing feedrate.

``` cpp
// Proportion of homing speed for the first approach when probing
#define Z_PROBE_FACTOR_FAST 1
// Proportion of homing speed for the second approach when probing
#define Z_PROBE_FACTOR_SLOW 0.5
```

AnHardt commented:
Changing to factor here would (for me) be repeating the same error.
With absolute speeds we have the full flexibility to use a factor, in the config., or not.
I admit using a variable here is a somewhat unusual construct. Do you have a better idea to get `homing_feedrate_mm_m[Z_AXIS]` into the config. (array defined)

Why `homing_feedrate_mm_m` at all? To give a speed that should work for most systems. But in principe they should be independent.  At least that was my takeaway from #3975.

thinkyhead commented:
I understand the point. But we cannot have (any more) configuration options that use named variables from the source code. So it has to be an absolute value or a factor related to some other reference speed.

Too bad the feedrates aren't all specified in one place.
Meanwhile, we could split them up a bit, like so…

``` cpp
#define HOMING_FEEDRATE_XY 50*60
#define HOMING_FEEDRATE_Z  4*60
#define HOMING_FEEDRATE {HOMING_FEEDRATE_XY, HOMING_FEEDRATE_XY, HOMING_FEEDRATE_Z, 0}  // set the homing speeds (mm/min)

. . .

#define Z_PROBE_SPEED_FAST HOMING_FEEDRATE_Z
#define Z_PROBE_SPEED_SLOW HOMING_FEEDRATE_Z / 2
```

AnHardt committed:
Unify run_z_probe

Unify run_z_probe
Add double touch for DELTAs.
Introduce Z_PROBE_SPEED_FAST and Z_PROBE_SPEED_SLOW
  defaulting to homing_feedrate_mm_m[Z_AXIS] and homing_feedrate_mm_m[Z_AXIS]/2

AnHardt commented:
Looks good to me. But i already hear Roxy moaning about to big changes in the configs.

The structure and order of the configs needs some more discussion, i think.
At the moment, for me, Configuration.h looks like after an earthquake.
With too much white space. Too long. Not in the order i would configure it for a new machine. ...
This should have a # in issues.

thinkyhead commented:
> `Configuration.h` looks like after an earthquake

Too many options? All out of order? Organization would be good!

thinkyhead commented:
Ready to merge #4356
